core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,1024 @@
+    /// This should be called in advance
+    /// in case we are really late getting shreds
+    fn prepare_to_receive(&mut self, slot: Slot) -> Result<(), Slot> {
+        trace!(
+            "{}: preparing to receive for slot {slot}",
+            self.cluster_info.id()
+        );
+        let staked_nodes = self.epoch_specs.current_epoch_staked_nodes();
+
+        let mut state = get_state_for_slot(&self.state, slot).lock().unwrap();
+        if state.current_slot != 0 {
+            return Err(state.current_slot);
+        }
+        state.current_slot = slot;
+        state.current_slot_start = Instant::now();
+        for (peer, &stake) in staked_nodes.iter() {
+            let Some(ag_addr) = self
+                .cluster_info
+                .lookup_contact_info(peer, |ci| ci.alpenglow())
+                .flatten()
+            else {
+                continue;
+            };
+            state.peers.insert(
+                *peer,
+                PeerData {
+                    stake,
+                    address: ag_addr,
+                    relative_toa: [None; NUM_VOTOR_TYPES],
+                },
+            );
+            state.total_staked += stake;
+        }
+        trace!(
+            "Prepared for slot {slot}, total stake is {}",
+            state.total_staked
+        );
+        Ok(())
+    }
+
+    /// Collects votes and changes states
+    fn listener_thread(
+        self_state: Arc<StateArray>,
+        should_exit: Arc<AtomicBool>,
+        my_id: Pubkey,
+        socket: Arc<UdpSocket>,
+        command_sender: Sender<SendCommand>,
+    ) {
+        socket
+            .set_read_timeout(Some(Duration::from_secs(1)))
+            .unwrap();
+        trace!("Listener thread started");
+        // Set aside enough space to fetch multiple packets from the kernel per syscall
+        let mut packets: Vec<Packet> = vec![Packet::default(); 1024];
+        loop {
+            // must wipe all Meta records to reuse the buffer
+            for p in packets.iter_mut() {
+                *p.meta_mut() = Meta::default();
+            }
+
+            if should_exit.load(Ordering::Relaxed) {
+                return;
+            }
+            // recv_mmsg should timeout in 1 second
+            let n = match recv_mmsg(&socket, &mut packets) {
+                // we may have received no packets, in this case we can safely skip the rest
+                Ok(0) => continue,
+                Ok(n) => n,
+                Err(e) => {
+                    match e.kind() {
+                        std::io::ErrorKind::TimedOut | std::io::ErrorKind::WouldBlock => {
+                            0 // no packets received
+                        }
+                        _ => {
+                            error!("Got error {:?} in mock alpenglow RX socket operation, exiting thread", e.raw_os_error());
+                            return;
+                        }
+                    }
+                }
+            };
+
```

Comment:
Just double checking, is this essentially the same receive logic as we use in FetchStage/any reason not to use that? Trying to replicate this as close as possible

Reply:
FetchStage is a wrapper around streamer, which ultimately calls recv_mmsg. Here we are just bypassing the 2 layers we do not need and call recv_mmsg directly.